src/LockTable.cc

```diff
@@ -0,0 +1,225 @@
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL AUTHORS BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include "LockTable.h"
+#include "BitOps.h"
+#include "Memory.h"
+
+namespace RAMCloud {
+
+/**
+ * Constructor for LockTable.
+ *
+ * \param numBuckets
+ *      Locks in a LockTable are stored in one or more buckets.  This parameter
+ *      sets the number of buckets that should be used for this new LockTable.
+ *      This value should be a power of two.
+ *
+ * \throw Exception
+ *      An exception is thrown if the resulting number of buckets is 0.
+ */
+LockTable::LockTable(uint64_t numBuckets)
+    : bucketIndexHashMask(BitOps::powerOfTwoLessOrEqual(numBuckets) - 1)
+    , buckets()
+{
+    if (numBuckets != (bucketIndexHashMask + 1)) {
+        RAMCLOUD_LOG(DEBUG,
+                     "LockTable truncated to %lu buckets "
+                     "(nearest power of two)",
+                     (bucketIndexHashMask + 1));
+    }
+
+    if ((bucketIndexHashMask + 1) == 0)
+        throw Exception(HERE, "LockTable numBuckets == 0?!");
+
+    void *buf  = Memory::xmemalign(
+            HERE,
+            sizeof(CacheLine),
+            (bucketIndexHashMask + 1) * sizeof(CacheLine));
+    memset(buf, 0, (bucketIndexHashMask + 1) * sizeof(CacheLine));
+    buckets = static_cast<CacheLine *>(buf);
+}
+
+/**
+ * Destructor for LockTable.
+ */
+LockTable::~LockTable() {
+    // Free any extra xmalloced CacheLines.
+    for (uint64_t i = 0; i < bucketIndexHashMask + 1; i++) {
+        CacheLine* cacheLine = buckets[i].next;
+        while (cacheLine != NULL) {
+            CacheLine* next = cacheLine->next;
+            free(cacheLine);
+            cacheLine = next;
+        }
+    }
+    free(buckets);
+}
+
+/**
+ * Blocks until the lock with the provided lockId can be acquired.
+ *
+ * \param lockId
+ *      ID of lock to be acquired.  ID must not be zero.
+ */
+void
+LockTable::acquireLock(uint64_t lockId)
+{
+    while (!tryAcquireLock(lockId))
+        continue;
+}
+
+/**
+ * Check if lock with the provided lockId is currently acquired.
+ *
+ * \param lockId
+ *      ID of lock to be acquired.  ID must not be zero.
+ * \return
+ *      TRUE if the lock is currently acquired, FALSE otherwise.
+ */
+bool
+LockTable::isLockAcquired(uint64_t lockId)
+{
+    // Find the right bucket.
+    uint64_t bucketIndex = (lockId & bucketIndexHashMask);
+    CacheLine* cacheLine = &buckets[bucketIndex];
+
+    // Lock the bucket
+    BucketLock* bucketLock =
+            reinterpret_cast<BucketLock*>(&cacheLine->entries[0]);
+    std::lock_guard<BucketLock> lock(*bucketLock);
```

Comment:
The code above is repeated in many methods; is there a simple way this could be extracted into a separate method (inline, for best performance)?